src/Native/System.Net.Security.Native/pal_ntlmapi.cpp

```diff
@@ -0,0 +1,266 @@
+{
+    assert(type2 != nullptr);
+    heim_ntlm_free_type2(type2);
+    delete type2;
+}
+
+extern "C" int32_t NetSecurity_HeimNtlmNtKey(const char* password, ntlm_buf** outBufferHandle, int32_t* outLength)
+{
+    assert(outBufferHandle != nullptr);
+    *outBufferHandle = new ntlm_buf();
+    return NetSecurity_SetBufferLength(heim_ntlm_nt_key(password, *outBufferHandle), outBufferHandle, outLength);
+}
+
+extern "C" int32_t NetSecurity_HeimNtlmCalculateResponse(int32_t isLM,
+                                                         const ntlm_buf* key,
+                                                         ntlm_type2* type2,
+                                                         char* username,
+                                                         char* target,
+                                                         uint8_t* baseSessionKey,
+                                                         int32_t baseSessionKeyLen,
+                                                         ntlm_buf** outBufferHandle,
+                                                         int32_t* outLength)
+{
+    // reference doc: http://msdn.microsoft.com/en-us/library/cc236700.aspx
+    assert(baseSessionKeyLen == MD5_DIGEST_LENGTH);
+    assert(isLM == 0 || isLM == 1);
+    assert(type2 != nullptr);
+    assert(key != nullptr);
+    assert(outBufferHandle != nullptr);
+    *outBufferHandle = new ntlm_buf();
+    if (isLM)
+    {
+        return NetSecurity_SetBufferLength(
+            heim_ntlm_calculate_lm2(
+                key->data, key->length, username, target, type2->challenge, baseSessionKey, *outBufferHandle),
+            outBufferHandle,
+            outLength);
+    }
+    else
+    {
+        if (type2->targetinfo.length == 0)
+        {
+            return NetSecurity_SetBufferLength(
+                heim_ntlm_calculate_ntlm1(key->data, key->length, type2->challenge, *outBufferHandle),
+                outBufferHandle,
+                outLength);
+        }
+        else
+        {
+            return NetSecurity_SetBufferLength(heim_ntlm_calculate_ntlm2(key->data,
+                                                                         key->length,
+                                                                         username,
+                                                                         target,
+                                                                         type2->challenge,
+                                                                         &type2->targetinfo,
+                                                                         baseSessionKey,
+                                                                         *outBufferHandle),
+                                               outBufferHandle,
+                                               outLength);
+        }
+    }
+}
+
+static uint8_t* NetSecurity_HMACDigest(uint8_t* key, int32_t keylen, void* input, size_t inputlen)
```

Comment:
These started out as static since they are really specific for heim_ntlm_build_ntlm2_master which is currently broken in libheimntlm. But I think there is value in making them extern and using them from managed code. @kapilash , in that case you wouldn't need to do the IncrementalHash changes either